operator/internal/handlers/lokistack_create_or_update.go

```diff
@@ -414,6 +221,252 @@ func CreateOrUpdateLokiStack(
+	if err != nil {
+		log.Error(err, "error in getting tenant secret data")
+	}
+
+	tenants = manifests.Tenants{
+		Secrets: secrets,
+		Configs: configs,
+	}
+
+	return baseDomain, tenants, nil
+}
+
+func getRulerOptions(
+	ctx context.Context,
+	log logr.Logger,
+	req ctrl.Request,
+	k k8s.Client,
+	stack *lokiv1.LokiStack,
+) ([]lokiv1.AlertingRule, []lokiv1.RecordingRule, manifests.Ruler, manifestsocp.Options, error) {
+	var (
+		err            error
+		alertingRules  []lokiv1.AlertingRule
+		recordingRules []lokiv1.RecordingRule
+		rulerConfig    *lokiv1.RulerConfigSpec
+		rulerSecret    *manifests.RulerSecret
+		ruler          manifests.Ruler
+		ocpOpts        manifestsocp.Options
+	)
+
+	if stack.Spec.Rules == nil || !stack.Spec.Rules.Enabled {
+		// Clean up ruler resources
+		err = rules.RemoveRulesConfigMap(ctx, req, k)
+		if err != nil {
+			log.Error(err, "failed to remove rules ConfigMap")
+			return nil, nil, ruler, ocpOpts, err
+		}
+
+		err = rules.RemoveRuler(ctx, req, k)
+		if err != nil {
+			log.Error(err, "failed to remove ruler StatefulSet")
+			return nil, nil, ruler, ocpOpts, err
+		}
+
+		return nil, nil, ruler, ocpOpts, nil
+	}
+
+	alertingRules, recordingRules, err = rules.List(ctx, k, req.Namespace, stack.Spec.Rules)
+	if err != nil {
+		log.Error(err, "failed to lookup rules", "spec", stack.Spec.Rules)
+	}
+
+	rulerConfig, err = rules.GetRulerConfig(ctx, k, req)
+	if err != nil {
+		log.Error(err, "failed to lookup ruler config", "key", req.NamespacedName)
+	}
+
+	if rulerConfig != nil && rulerConfig.RemoteWriteSpec != nil && rulerConfig.RemoteWriteSpec.ClientSpec != nil {
+		var rs corev1.Secret
+		key := client.ObjectKey{Name: rulerConfig.RemoteWriteSpec.ClientSpec.AuthorizationSecretName, Namespace: stack.Namespace}
+		if err = k.Get(ctx, key, &rs); err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, nil, ruler, ocpOpts, &status.DegradedError{
+					Message: "Missing ruler remote write authorization secret",
+					Reason:  lokiv1.ReasonMissingRulerSecret,
+					Requeue: false,
+				}
+			}
+			return nil, nil, ruler, ocpOpts, kverrors.Wrap(err, "failed to lookup lokistack ruler secret", "name", key)
+		}
+
+		rulerSecret, err = rules.ExtractRulerSecret(&rs, rulerConfig.RemoteWriteSpec.ClientSpec.AuthorizationType)
+		if err != nil {
+			return nil, nil, ruler, ocpOpts, &status.DegradedError{
+				Message: "Invalid ruler remote write authorization secret contents",
+				Reason:  lokiv1.ReasonInvalidRulerSecret,
+				Requeue: false,
+			}
+		}
+	}
+
+	ocpAmEnabled, err := openshift.AlertManagerSVCExists(ctx, stack.Spec, k)
+	if err != nil {
+		log.Error(err, "failed to check OCP AlertManager")
+		return nil, nil, ruler, ocpOpts, err
+	}
+
+	ocpUWAmEnabled, err := openshift.UserWorkloadAlertManagerSVCExists(ctx, stack.Spec, k)
+	if err != nil {
+		log.Error(err, "failed to check OCP User Workload AlertManager")
+		return nil, nil, ruler, ocpOpts, err
+	}
```

Comment:
Minor comment, it looks weird that the RulerOptions generate the OCP Options, although they are related, however, I also have a hard time justifying a function with only these two steps. Maybe in the future, it will become more evident